### DISCUSSION TITLE
Error server.on

### DISCUSSION THREAD

**Discussion by aber68** [Sloeber/arduino-eclipse-plugin#1648] [General]
hello, why give **"server.on("/", HTTP_GET, handleRoot);"** an error: "Invalid Arguements" in a WebServer sketch? br andreas

![grafik](https://github.com/Sloeber/arduino-eclipse-plugin/assets/38425464/152eeaa4-2111-4970-a1ed-8a1aa7044a58)

ps: compiling works but this errors above show up

**Comment by jantje** [comment]
I have no clue why but this is a typical code analyzer problem.
There are many in CDT. The more advanced C code you use the more likely you will meet them.

**Reply by thatguy-jaysenodell** [reply]
I chased a similar issue with esp8266 and WiFiUDP libraries. Effectively the ArduinoOTA lib was loading WIFI -> WiFiUDP and overriding ESP8266WiFi internal WiFiUDP. I had to fight with included libraries to only include the actual libraries I needed vs the automatic linking of libs into the project. The workflow that I have managed to successully use over multiple efforts in the last couple days. In your include ordering you need use the following structure: 
1. start with the bast project includes (arduino.h, board/chip/platform).
2. add other hardware components but NOT the software (esp8266wifi vs wifi+spi).
3. add low level sofware to drive hardware interfaces only as you find you absolutely need them (wire, spi, etc).
4. add all other helper libs (esp8266webserver, littlefs/spiffs, AsycHttpServer, etc)
At each step, save your .h or .cpp/.ino and remove lib that you know are not needed. For example, when I add ESP8266WiFi.h in step 2, then add Arduino.h in step 4, Sloeber will add WiFi... I need to remove WiFi lib in the library manager. 

Notice I'm esp8266 heavy here, but the ESP libs tend to redefine a pile of "standard" constructors (WiFiUDP and HTTPServer for example). If the example you are following is written for an ESP module (or some other "do it my own way" implementation) the include order may be tripping you up. 

That's what I found anyway. 

There is a way to turn off autolinking of libs, but I haven't looked into it because I'm lazy and wanted to get OTA uploads for filesystems working. It seems to me that autolinking is a great feature for the most part. Not sure it's one I'm eager to disable. 

Evidence on the conflict for WiFiUDP between ESP8266WiFi and standard WiFi...
```
jaysen@mba13 arduinoPlugin % pwd; find . -name WiFiUdp.h
~/Applications/eclipse/cpp-2024-06/Eclipse.app/Contents/Eclipse/arduinoPlugin
./libraries/WiFi/1.2.7/src/WiFiUdp.h
./libraries/WiFiNINA/1.8.14/src/WiFiUdp.h
./packages/esp8266/hardware/esp8266/3.1.2/libraries/ESP8266WiFi/src/WiFiUdp.h
./packages/esp8266/hardware/esp8266/3.0.2/libraries/ESP8266WiFi/src/WiFiUdp.h
```

**Note**
I don't think Sloeber is wrong here. ArduinoOTA includes WiFi.h which includes local WiFiUdp.h. The library resolver should include the WiFi lib. Then under compile it's eplicitly adding the WiFi lib before the ESP8266WiFi lib. The problem is that gcc is much picker under Sloeber than it is under Arduino IDE. This isn't bad. It's just different. as noted by @jantje, this is just one difficulty with using advanced tooling.
